Use sbt for PR validation [ci: last-only]

szeiger opened the pull request:
Picking up the work on #4933

szeiger commented:
Ah, we still need `pull-binary-libs` call for partest to fetch some JARs. The larger jsoup JAR could probably be resolved from Maven Central and the bunch of small JARs added directly to the git repo.

But this doesn't explain the test failure. The `pull-binary-libs` call is still there and you can clearly see from the log that it resolves the files, but none of them end up on the scalac classpath for partest. I was unable to reproduce this issue locally.

retronym commented:
@szeiger Related discussion: https://issues.scala-lang.org/browse/SI-8379

retronym commented:
>  bunch of small JARs added directly to the git repo.

We could also rewrite the tests to generate the necessary bytecode programmatically. We can use ASMifier to reify the ASM code that would create a given class, and then whittle that down the miminum needed to provoke the original bug.

szeiger commented:
It looks like we have our first successful PR validation build on sbt. Remaining issues:

- [ ] Run scaladoc tests from sbt
- [ ] Port OSGi tests to sbt
- [ ] Put the build artifacts in the same location as the ant build or change the Jenkins configuration:
```
Archiving artifacts
WARN: No artifacts found that match the file pattern "build/junit/TEST-*,build/osgi/TEST-*,hs_err_*.log". Configuration error?
WARN: ‘build/junit/TEST-*’ doesn’t match anything: ‘build’ exists but not ‘build/junit/TEST-*’
```

The following are non-issues (at least for this PR) IMHO:
- Respect `scalac.args.optimise` - The PR validation build is always with optimization. If we need other builds without, they get different setup commands (in `project/ScriptCommands.scala`) where optimization can be left off.
- Do not require `pull-binary-libs` for PR validation (or anything else) - We should do this but it's not a requirement for switching from ant to sbt, so not in scope for this PR.

InTheNow commented:
re scaladoc. We have issues generating scaladoc from sbt in cats. I know the issue/solution but have not PR'd anything- ref chat https://gitter.im/sbt/sbt?at=574b4bde10f0fed86f49499c

In case you hit any issues before I PR anything, please shout.

szeiger commented:
@InTheNow Scaladoc generation has been working in the Scala build for months. The remaining issue here is about running tests on the scaladoc tool (using partest from sbt)

szeiger commented:
Marking "Put the build artifacts in the same location as the ant build or change the Jenkins configuration" as solved. Some build artifacts were produced by the ant build and then archived by Jenkins because they contain additional details about build errors. The sbt build logs everything to the console instead, so these files should not be necessary anymore. The warning can be safely ignored and the Jenkins configuration changed after this PR has been merged.

szeiger commented:
Ready for review.

adriaanm commented on build.sbt:
```diff
@@ -326,6 +321,11 @@ def regexFileFilter(s: String): FileFilter = new FileFilter {
 }
 
+// This project provides the STARR scalaInstance for bootstrapping
+lazy val bootstrap = (project in file("target/bootstrap")).settings(
```
neat!

lrytz replied:
i wonder about the various ways to create `File`s that we see in `build.sbt`:
- `file("target/...")`
- `file(".") / "target" / ...`
- `(baseDirectory in ThisBuild).value / "target" / ...`
- `buildDirectory.value / "quick" / ...`

does it matter, should we unify / clean up a bit?

szeiger replied:
We should be able to get rid of many of those when we move to a more standard sbt layout after ant is gone. (The first 3 all do the same, the last one is different)

lrytz replied:
here: if you tell me we're fine for now then that's fine with me :)

adriaanm commented:
I would really like to reduce the additional noise in the output. Maybe suppress [info]-level output? I know it's annoying to fix, but if we postpone it I'm afraid we'll just end up living with it.

retronym commented:
I just tried to use the native SBT import in IntelliJ to see if the changes in this PR had been enough to get it working. (It wasn't working before because of our custom `scalaInstance`.)

 It does successfully complete, but ends up with the wrong directory mappings:

![image](https://cloud.githubusercontent.com/assets/65551/15796572/4ab5128e-2a43-11e6-8261-406ed900afe5.png)

Not a blocker for this PR, just a pity.

retronym commented:
> It does successfully complete, but ends up with the wrong directory mappings:

Oh, that's a side effect of a change I made previously:

```
  // When we fork subprocesses, use the base directory as the working directory.
  // This enables `sbt> partest test/files/run/t1.scala` or `sbt> scalac sandbox/test.scala`
  baseDirectory in Compile := (baseDirectory in ThisBuild).value,
  baseDirectory in Test := (baseDirectory in ThisBuild).value,
```

IntelliJ [looks up `compile:baseDirectory`](https://github.com/JetBrains/sbt-structure/blob/a8075f0ed59b036a5e55782a30bffabad5397766/extractor/src/main/scala/org/jetbrains/sbt/extractors/ProjectExtractor.scala#L112) rather than `baseDirectory` to use as the content root of the module.

szeiger commented:
@adriaanm What noise are you referring to specifically? The OSGi tests are noisy but I didn't bother trying to change that (not sure that I could - I only managed to turn some of that noise off in Slick's OSGi tests) because you don't normally run them manually and the console output replaces the JUnit XML logs from the ant build.

adriaanm commented:
Most lines that start with `[info]`. For example, the ones about `resolving ...`. The JUnit tests also output more than necessary.

lrytz commented on build.sbt:
```diff
@@ -616,18 +667,17 @@ lazy val test = project
-          def annotationName = "partest"
-        }, true, Array())
-     )
+    testOptions in IntegrationTest += Tests.Argument("-Dpartest.scalac_opts=" + (scalacOptions in Compile).value.mkString(" ")),
```
it looks like space separation might not work if there are multiple scalac options

```
-Dpartest.scalac_opts=-optimize -deprecation
```

maybe surround the options by single quotes? just guessing..

szeiger replied:
No, there's no parsing of quoted arguments happening anywhere. This is a single `Argument` for partest which does a `.split(' ')` on the RHS of the `-D` def to get the scalac args. There is currently no way use spaces inside of scalac args but that's a limitation of partest (and is also present when launching it from ant or the command line).

lrytz replied:
if you tell me we're fine for now then that's fine with me :)

lrytz replied:
oh, wrong spot for this comment.

lrytz replied:
OK, thanks!

lrytz commented on build.sbt:
```diff
@@ -701,16 +751,18 @@ lazy val scalaDist = Project("scala-dist", file(".") / "target" / "scala-dist-di
-)
+    Quiet.silenceIvyUpdateInfoLogging,
+    antStyle := false,
+    incOptions := incOptions.value.withNameHashing(!antStyle.value).withAntStyle(antStyle.value)
```
I used to have a separate `local.sbt` (which I keep in my global `~/.gitignore`) with the following line:

```
(incOptions in ThisBuild) := (incOptions in ThisBuild).value.withNameHashing(false).withAntStyle(true)
```

This gives an error now:

```
> compile
...
[error] (library/*:incOptions) java.lang.AssertionError: assertion failed: Name hashing and Ant-style cannot be enabled at the same time.
...
```

I tried putting `antStyle := true` in my `local.sbt`, but that doesn't work (`local.sbt:1: error: not found: value antStyle`). What is the correct way to enable it? I wouldn't like to type `set antStyle := true` each time I start sbt.

dwijnand replied:
One way is by re-defining `antStyle` in `local.sbt`, sbt will merge them if they have the same type.

lrytz replied:
oh, my mistake, i just did that on current 2.12.x, let me try again what's the state after this PR!

lrytz replied:
wrong spot again, sorry..!!!!

lrytz replied:
that works, thanks!

szeiger replied:
I've asked the sbt guys for help with this because I couldn't make it work. If there's no cleaner solution, I could move the definition into an object (which can be referenced from all `.sbt` files) and add an alias in `build.sbt` to make it available with an unqualified name on the command line.

dwijnand replied:
An alternative is to define a local AutoPlugin just for the purpose of defining auto-imported keys: https://github.com/scala-js/scala-js/blob/af7c61d43bc0f9395fa6aec2ca2aecfcb0883636/project/Build.scala#L42-L50

lrytz commented on build.sbt:
```diff
@@ -616,18 +667,17 @@ lazy val test = project
     fork in IntegrationTest := true,
     javaOptions in IntegrationTest += "-Xmx1G",
     testFrameworks += new TestFramework("scala.tools.partest.sbt.Framework"),
-    testOptions in IntegrationTest += Tests.Setup( () => root.base.getAbsolutePath + "/pull-binary-libs.sh" ! ),
```
so when I run

```
sbt
> partest test/files/pos/cycle-jsoup.scala
```

on a fresh checkout (or after `git clean -dfx`), I get:

```
...
[info] Packaging /Users/luc/scala/scala4/build/pack/lib/scala-partest-javaagent.jar ...
[info] Done packaging.
...
Resolving [346d3dff4088839d6b4d163efa2892124039d216/test/files/lib/jsoup-1.3.1.jar]
...
Partest version:
Compiler under test: $baseDir/compiler
Scala version is:    Scala compiler version 2.12.0-20160606-115156-79e7334 -- Copyright 2002-2016, LAMP/EPFL
Scalac options are:
Compilation Path:    /Users/luc/scala/scala4/target/test/it-classes:$baseDir/test:$baseDir/compiler:$baseDir/library:$baseDir/reflect:$baseDir/interactive:$baseDir/repl-jline-embedded:$baseDir/repl-jline:$baseDir/repl:$baseDir/scalap:$baseDir/partest-extras:/Users/luc/scala/scala4/build/pack/lib/scala-partest-javaagent.jar:$baseDir/scaladoc:/Users/luc/scala/scala4/test/lib:/Users/luc/.ivy2/cache/org.apache.ant/ant/jars/ant-1.9.4.jar:/Users/luc/.ivy2/cache/org.apache.ant/ant-launcher/jars/ant-launcher-1.9.4.jar:/Users/luc/.ivy2/cache/org.scala-lang.modules/scala-asm/bundles/scala-asm-5.0.4-scala-3.jar:/Users/luc/.ivy2/cache/org.scala-lang.modules/scala-xml_2.12.0-M4/bundles/scala-xml_2.12.0-M4-1.0.5.jar:/Users/luc/.ivy2/cache/org.scala-lang.modules/scala-parser-combinators_2.12.0-M4/bundles/scala-parser-combinators_2.12.0-M4-1.0.4.jar:/Users/luc/.ivy2/cache/jline/jline/jars/jline-2.14.1.jar:/Users/luc/.ivy2/cache/org.scala-lang.modules/scala-partest_2.12.0-M4/jars/scala-partest_2.12.0-M4-1.0.14.jar:/Users/luc/.ivy2/cache/com.googlecode.java-diff-utils/diffutils/jars/diffutils-1.3.0.jar:/Users/luc/.ivy2/cache/org.scala-sbt/test-interface/jars/test-interface-1.0.jar:/Users/luc/.ivy2/cache/org.scalacheck/scalacheck_2.12.0-M4/jars/scalacheck_2.12.0-M4-1.11.6.jar:/Users/luc/.sbt/boot/scala-2.10.6/org.scala-sbt/sbt/0.13.11/test-agent-0.13.11.jar:/Users/luc/.sbt/boot/scala-2.10.6/org.scala-sbt/sbt/0.13.11/test-interface-1.0.jar
Java binaries in:    /Library/Java/JavaVirtualMachines/jdk1.8.0_77.jdk/Contents/Home/jre/bin
Java runtime is:     Java HotSpot(TM) 64-Bit Server VM (build 25.77-b03, mixed mode)
Java options are:    -Xmx1024M -Xms64M -XX:MaxPermSize=128M
baseDir:             /Users/luc/scala/scala4/build/quick/classes
sourceDir:           /Users/luc/scala/scala4/test/files

Selected 1 tests drawn from specified tests

# starting 1 test in pos
!! 1 - pos/cycle-jsoup.scala                     [compilation failed]
```

with

```
cycle-jsoup.scala:3: error: object jsoup is not a member of package org
    org.jsoup.Jsoup.parse(null: java.net.URL, 3000)
        ^
one error found
```

If I then run the same command `partest test/files/pos/cycle-jsoup.scala` again in the same sbt shell process (no restart), the test passes:

```
> partest test/files/pos/cycle-jsoup.scala --show-diff
[info] Packaging /Users/luc/scala/scala4/build/pack/lib/scala-partest-javaagent.jar ...
[info] Done packaging.
Partest version:
Compiler under test: $baseDir/compiler
Scala version is:    Scala compiler version 2.12.0-20160606-115156-79e7334 -- Copyright 2002-2016, LAMP/EPFL
Scalac options are:
Compilation Path:    /Users/luc/scala/scala4/target/test/it-classes:$baseDir/test:$baseDir/compiler:$baseDir/library:$baseDir/reflect:$baseDir/interactive:$baseDir/repl-jline-embedded:$baseDir/repl-jline:$baseDir/repl:$baseDir/scalap:$baseDir/partest-extras:/Users/luc/scala/scala4/build/pack/lib/scala-partest-javaagent.jar:$baseDir/scaladoc:$sourceDir/lib/annotations.jar:$sourceDir/lib/enums.jar:$sourceDir/lib/genericNest.jar:$sourceDir/lib/jsoup-1.3.1.jar:$sourceDir/lib/macro210.jar:$sourceDir/lib/methvsfield.jar:$sourceDir/lib/nest.jar:/Users/luc/scala/scala4/test/lib:/Users/luc/.ivy2/cache/org.apache.ant/ant/jars/ant-1.9.4.jar:/Users/luc/.ivy2/cache/org.apache.ant/ant-launcher/jars/ant-launcher-1.9.4.jar:/Users/luc/.ivy2/cache/org.scala-lang.modules/scala-asm/bundles/scala-asm-5.0.4-scala-3.jar:/Users/luc/.ivy2/cache/org.scala-lang.modules/scala-xml_2.12.0-M4/bundles/scala-xml_2.12.0-M4-1.0.5.jar:/Users/luc/.ivy2/cache/org.scala-lang.modules/scala-parser-combinators_2.12.0-M4/bundles/scala-parser-combinators_2.12.0-M4-1.0.4.jar:/Users/luc/.ivy2/cache/jline/jline/jars/jline-2.14.1.jar:/Users/luc/.ivy2/cache/org.scala-lang.modules/scala-partest_2.12.0-M4/jars/scala-partest_2.12.0-M4-1.0.14.jar:/Users/luc/.ivy2/cache/com.googlecode.java-diff-utils/diffutils/jars/diffutils-1.3.0.jar:/Users/luc/.ivy2/cache/org.scala-sbt/test-interface/jars/test-interface-1.0.jar:/Users/luc/.ivy2/cache/org.scalacheck/scalacheck_2.12.0-M4/jars/scalacheck_2.12.0-M4-1.11.6.jar:/Users/luc/.sbt/boot/scala-2.10.6/org.scala-sbt/sbt/0.13.11/test-agent-0.13.11.jar:/Users/luc/.sbt/boot/scala-2.10.6/org.scala-sbt/sbt/0.13.11/test-interface-1.0.jar
Java binaries in:    /Library/Java/JavaVirtualMachines/jdk1.8.0_77.jdk/Contents/Home/jre/bin
Java runtime is:     Java HotSpot(TM) 64-Bit Server VM (build 25.77-b03, mixed mode)
Java options are:    -Xmx1024M -Xms64M -XX:MaxPermSize=128M
baseDir:             /Users/luc/scala/scala4/build/quick/classes
sourceDir:           /Users/luc/scala/scala4/test/files

Selected 1 tests drawn from specified tests

# starting 1 test in pos
ok 1 - pos/cycle-jsoup.scala
```

note that `jsoup-1.3.1.jar` is only on the "Compilation Path" in the second invocation.

lrytz replied:
here: oh, my mistake, i just did that on current 2.12.x, let me try again what's the state after this PR!

lrytz replied:
yeah, so as expected: after this PR, `partest test/files/pos/cycle-jsoup.scala` fails until `pull-binary-libs.sh` is run manually. i think we should find a way to fix this, it's too confusing for newcomers.

szeiger replied:
This is expected. You still need to run `pull-binary-libs.sh` initially. Note that even with the now removed line it used to fail on the first run because doing it in `Test.Setup` is too late (the classpath has already been built before the JARs are downloaded). Only subsequent runs would succeed.

But I don't really want to call it from the sbt build at all because that would require a Unix environment. Since most of the binaries are already resolved through Ivy, I hope that we can get of it entirely. JSoup is on Maven Central, the remaining binary files are about 10 kB in total. It would be nice to generate them on demand but the size is so low that we could just check them into git. Another option is to download them with Scala code in the sbt build.

lrytz replied:
OK, agree. could we maybe add an assertion somewhere in the sbt startup, require that `test/files/lib/annotations.jar` exists?

szeiger replied:
We could check for it in `Test.Setup` and log a warning if the JARs are missing. I don't think the whole build should fail. Only a handful of partest test cases still need `pull-binary-libs`. Everything else is resolved through Ivy in the sbt build.

lrytz commented on project/MiMa.scala:
```diff
@@ -0,0 +1,94 @@
+
+  lazy val settings =
+    Seq(
+      mimaOldVersion := VersionUtil.versionProps("scala.binary.version") + ".0",
```
not sure if using `scala.binary.version` is correct here, see its doc comment and its value in the 2.12.x branch: https://github.com/scala/scala/blob/2.12.x/versions.properties#L16-L22

the code here will only work after `.0` is released. also note that we'd like to enable mima once M5 is out and set M5 (or maybe later RC1) as the base version. in ant we used `2.11.0` directly, no?

related, what is the scenario to enable / disable mima during the milestone cycle? just don't invoke it from the CI bash scripts?

szeiger replied:
I assumed that MiMa would only be enabled after a .0 release. We can add a separate base version setting to the sbt build. It's also hardcoded in build.xml. The ant build reacts to a `test.bc.skip` setting but I don't know if it's actually used. I would change it in the CI script or build.sbt as required (especially if the base version is also set there; these two settings belong together)

lrytz replied:
agreed

lrytz commented on test/scaladoc/run/t8557.scala:
```diff
@@ -15,10 +17,22 @@ object Test extends ScaladocModelTest {
+  override def scaladocSettings = {
+    val samplePath = getClass.getClassLoader.getResource("scala/Function1.class").getPath.replace('\\', '/')
+    val scalaLibPath = if(samplePath.contains("!")) { // in scala-library.jar
+    val scalaLibUri = samplePath.split("!")(0)
```
nit: indent this line

szeiger replied:
Another victim of IntelliJ's broken formatter (doesn't indent the line after a `{` when followed by a line comment)

szeiger commented:
/rebuild

lrytz commented:
```
Caused by: sbt.ForkMain$ForkError: java.lang.RuntimeException: 'instrumented.jar' not found in '/home/jenkins/workspace/scala-2.11.x-validate-test/test/files/speclib'.
```

retronym commented:
> Most lines that start with [info]. For example, the ones about resolving .... The JUnit tests also output more than necessary.

@adriaanm I turned that off in https://github.com/scala/scala/pull/5152/files#diff-991b9473f196f6364820f0c3652eb3faR32 . Perhaps you are on a branch doesn't have that change it its hiatory yet?

retronym commented:
> Most lines that start with [info]. For example, the ones about resolving .... The JUnit tests also output more than necessary.

@adriaanm I turned that off in https://github.com/scala/scala/pull/5152/files#diff-991b9473f196f6364820f0c3652eb3faR32 . Perhaps you are on a branch doesn't have that change it its hiatory yet?

szeiger commented:
@retronym I've just pushed another commit with a workaround for your scalacheck bootstrapping problem (provided that it passes CI and doesn't cause new problems).

adriaanm commented:
> Perhaps you are on a branch doesn't have that change it its hiatory yet?

I was looking at the output of this PR, so I guess this that'll take care of itself when we merge.

szeiger commented:
No, it's in the history and I see the code. It may be scoped too narrowly to catch everything. I'll investigate. (BTW, noise has already been considerably reduced by running everything from a single sbt invocation; now you only get the resolver spam once at the beginning)

szeiger commented:
Everything should be fixed now.

adriaanm commented:
Fix all the things!! :party:

szeiger committed:
Use sbt for PR validation

- Support directories in `-doc-external-doc`: It is documented as
  accepting a “classpath_entry_path” for the keys but this only worked
  for JARs and not for individual class files. When checking for
  external-doc mappings for a Symbol, we now find the root directory
  relative to a class file instead of using the full class file path.
  The corresponding tests for SI-191 and SI8557 are also fixed to
  support individual class files instead of JARs in partest. This is
  required for the sbt build which runs partest on “quick” instead of
  “pack”.

- Fix version and repository handling for bootstrapping. The bootstrap
  `scalaInstance` can now be resolved from any repository added to the
  project (not just the bootstrap repositories) by using a different
  workaround for https://github.com/sbt/sbt/issues/1872.

- Workaround for https://github.com/sbt/sbt/issues/2640 (putting the
  wrong `scalaInstance` on partest’s classpath). The required
  `ScalaInstance` constructor is deprecated, so we have to disable
  deprecation warnings and fatal warnings until there is a better fix.

- Add MiMa to the sbt build (port of the old `test.bc` ant task). The
  sbt-mima plugin doesn’t have all the features we need, so we do it
  manually in a similar way to what the plugin does. Checks are done
  in both directions for the `library` and `compiler` projects. The
  base version has to be set in `build.sbt`. When set to `None`, MiMa
  checks are skipped. MiMa checks are run sequentially to avoid spurious
  errors (see https://github.com/typesafehub/migration-manager/issues/115).

- Port the OSGi tests to the sbt build. The set of JARs that gets copied
  into build/osgi as bundles is a bit different from the ant build. We
  omit the source JARs but add additional modules that are part of the
  Scala distribution, which seems more correct.

- Get rid up `pull-binary-libs.sh` for the sbt build. Add artifacts are
  resolved from the special bootstrap repository through Ivy. The
  special `code.jar` and `instrumented.jar` artifacts are copied to the
  location where partest expects them (because these paths are hardcoded
  in partest). Other extra JARs for partest in `test/files/lib` are
  referenced directly from the Ivy cache.

- Move common settings that should be available with unqualified names
  in local `.sbt` files and on the command line into an auto-plugin.

- Add an `antStyle` setting to sbt to allow users to easily enable
  ant-style incremental compilation instead of sbt’s standard name
  hashing with `set antStyle := true`.

- Disable verbose `info`-level logging during sbt startup for both,
  `validate/test` and `validate/publish-core` jobs. Update logging is
  no longer disabled when running locally (where it is useful and
  doesn’t generate excessive output).

- Pass optimization flags for scalac down to partest, using the new
  partest version 1.0.15\6.

- Call the new sbt-based PR validation from `scripts/jobs/validate/test`.

- Disable the tests `run/t7843-jsr223-service` and `run/t7933` from
  https://github.com/scala/scala/pull/4959 for now. We need to set up
  a new test project (either partest or junit) that can run them on a
  packaged version of Scala, or possibly move them into a separate
  project that would naturally run from a packaged Scala as part of the
  community build.

szeiger commented:
Squashed, rebased and switched to partest 1.0.15

dwijnand commented on project/MiMa.scala:
```diff
@@ -0,0 +1,94 @@
+// It would be nice to use sbt-mima-plugin here, but the plugin is missing
+// at least two features we need:
+// * ability to run MiMa twice, swapping `curr` and `prev`, to detect
+//   both forwards and backwards incompatibilities
```
I think this isn't true as of 0.1.9: https://github.com/typesafehub/migration-manager/commit/2844ffa48b6d2255aa64bd687703aec21dadd55e

dwijnand commented on project/MiMa.scala:
```diff
@@ -0,0 +1,94 @@
+// at least two features we need:
+// * ability to run MiMa twice, swapping `curr` and `prev`, to detect
+//   both forwards and backwards incompatibilities
+// * ability to pass a filter file
```
I think this is tracked in https://github.com/typesafehub/migration-manager/issues/102. Perhaps it would be useful to add that as a link in the comment.

szeiger commented:
Offending jarlister tests disabled, updated to a new partest 1.0.16, MiMa comments added.

szeiger committed:
Remove bnd warnings from log output

lrytz commented:
LGTM! 🎈
